### PR TITLE
dbeaver/dbeaver#23135 Enhance the 'Can't create driver instance' error

### PR DESCRIPTION
**Pull request by serjiokov**
Enhance the 'Can't create driver instance' error

### CODE REVIEW

**Comment by ShadelessFox** [comment]
Please provide a screenshot

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/exec/DBCDriverException.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jkiss.dbeaver.model.exec;
+
+import org.jkiss.dbeaver.model.connection.DBPDriver;
+
+public class DBCDriverException extends DBCException {
+    private static final long serialVersionUID = 1L;
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/exec/DBCDriverException.java]: We don't serialize objects

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/exec/DBCDriverException.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    private String driverClassName;
+    private String driverFullName;
+    private DBPDriver driver;
+
+    public DBCDriverException(String message, String driverClassName, String driverFullName, DBPDriver driver,  Throwable e) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/exec/DBCDriverException.java]: Missing nullability annotations

File: plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/DriverEditDialog.java [hunk 3]
```diff
@@ -1020,12 +1054,22 @@ protected void createButtonsForButtonBar(Composite parent) {
...
+                    });
+                } else if (driver != null) {
+                    UIUtils.asyncExec(() -> {
+                        DriverEditDialog dialog = new DriverEditDialog(
+                            UIUtils.getActiveWorkbenchShell(),
+                            driver);
+                        dialog.open();
+                    });
+                }
+
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/DriverEditDialog.java]: Code duplication

File: product/debug/DBeaver.product.launch [hunk 4]
```diff
@@ -1,12 +1,7 @@
 <?xml version="1.0" encoding="UTF-8" standalone="no"?>
 <launchConfiguration type="org.eclipse.pde.ui.RuntimeWorkbench">
-    <stringAttribute key="application" value="org.jkiss.dbeaver.ui.app.standalone.standalone"/>
-    <stringAttribute key="configLocation" value="${workspace_loc}/.metadata/.plugins/org.eclipse.pde.core/DBeaver.product"/>
-    <stringAttribute key="location" value="${workspace_loc}/../runtime-DBeaver.product"/>
-    <stringAttribute key="product" value="org.jkiss.dbeaver.ui.app.standalone.product"/>
-    <stringAttribute key="productFile" value="\org.jkiss.dbeaver.ce.standalone\DBeaver.product"/>
-
     <booleanAttribute key="append.args" value="true"/>
+    <stringAttribute key="application" value="org.jkiss.dbeaver.ui.app.standalone.standalone"/>
```
Comment (ShadelessFox) [inline product/debug/DBeaver.product.launch]: Revert this file

File: plugins/org.jkiss.dbeaver.core/src/org/jkiss/dbeaver/core/CoreResources.properties [hunk 5]
```diff
@@ -595,4 +595,6 @@ popup_open_browser_open_external_browser = Open in external browser
 
 pref_page_accessibility_screen_reader_group_lbl = Screen reader
 pref_page_accessibility_screen_reader_msg = Screen reader
-pref_page_accessibility_screen_reader_description = Enable additional support of screen reader accessibility for database editors
\ No newline at end of file
+pref_page_accessibility_screen_reader_description = Enable additional support of screen reader accessibility for database editors
+
+initialization_driver_error_msg = Error of initialization driver name ''{0}'' class ''{1}''
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.core/src/org/jkiss/dbeaver/core/CoreResources.properties]:
```suggestion
initialization_driver_error_msg = Error initializing driver ''{0}'' (class: ''{1}'')
```

File: plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/driver/DriverDescriptor.java [hunk 6]
```diff
@@ -1274,8 +1275,14 @@ public void loadDriver(DBRProgressMonitor monitor, boolean forceReload)
                 try {
                     // Load driver classes into core module using plugin class loader
                     driverClass = Class.forName(driverClassName, true, classLoader);
-                } catch (Throwable ex) {
-                    throw new DBException("Error creating driver '" + getFullName() + "' instance.\nMost likely required jar files are missing.\nYou should configure jars in driver settings.\n\nReason: can't load driver class '" + driverClassName + "'", ex);
+                } catch (Throwable e) {
+                    throw new DBCDriverException("Driver libraries (.jars) is not supplied in the product delivery"
+                        + " or libraries (.jars) not found in path."
+                        + "\nDriver configuration is not valid",
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/driver/DriverDescriptor.java]: What's "product delivery"? This doesn't sound like a sentence written in English.

**Review by ShadelessFox — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/driver/DriverDescriptor.java (hunk 6 above)
Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/driver/DriverDescriptor.java]: Widely used term 
https://learningloop.io/glossary/product-delivery#:~:text=Product%20Delivery%20is%20a%20term,and%20quality%20of%20software%20development.

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/exec/DBCDriverException.java (hunk 1 above)
Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/exec/DBCDriverException.java]: We are, pls check the inheritance of DBException.

File: product/debug/DBeaver.product.launch (hunk 4 above)
Comment (serjiokov) [reply product/debug/DBeaver.product.launch]: Sorry will do

File: plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/DriverEditDialog.java (hunk 3 above)
Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/DriverEditDialog.java]: done

**Comment by serjiokov** [comment]
image

File: plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/driver/DriverDescriptor.java (hunk 6 above)
Comment (ShadelessFox) [reply plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/driver/DriverDescriptor.java]: I don't see this text on the screenshot below. Is it used somewhere else?

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/exec/DBCDriverException.java (hunk 1 above)
Comment (ShadelessFox) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/exec/DBCDriverException.java]: It's present here for some legacy reasons without any real usage.

File: plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/driver/DriverDescriptor.java (hunk 6 above)
Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/driver/DriverDescriptor.java]: Discussed internally to keep original text

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/exec/DBCDriverFilesMissingException.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+@SuppressWarnings("serial")
+public class DBCDriverFilesMissingException extends DBCException {
+
+    private DBPDriver driver;
+
+    public DBCDriverFilesMissingException(
+        @NotNull DBPDriver driver,
+        @NotNull Throwable e) {
+        super(DBCMessages.initialization_driver_error_details, e);
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/exec/DBCDriverFilesMissingException.java:32]:
```suggestion
        @NotNull Throwable e
    ) {
        super(DBCMessages.initialization_driver_error_details, e);
```

**Review by E1izabeth — APPROVED** [review]

**Review by LonwoLonwo — APPROVED** [review]

**Comment by uslss** [comment]
CE - no button is shown in case of test connection action on connection creation:
image

cloudbeaver - error message is shown as before - **will be fixed by CB team**
image

**Comment by uslss** [comment]
verified

**Pull request merged**